### PR TITLE
Implement internal support for synthesis modulo oracles

### PR DESCRIPTION
**Pull request by ajreynol**
SyMO is implemented as a preprocessor for SMT queries in the verification subsolver of SyGuS.

### CODE REVIEW

File: src/theory/quantifiers/sygus/synth_verify.h [hunk 1]
```diff
@@ -56,6 +56,12 @@ class SynthVerify : protected EnvObj
    * Preprocess query internal. This returns the rewritten form of query
    * and includes all relevant function definitions, i.e. those that occur
    * in query. These are added as top-level conjuncts to the returned formula.
+   *
+   * For each oracle function f in the query, we conjoin equalities f(c) = d
+   * where (c, d) is an I/O pair obtained for a call to a oracle. In contrast
+   * to the description Polgreen et al VMCAI 2022, the verification subcall
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/synth_verify.h:62]:
```suggestion
   * to the description in Polgreen et al VMCAI 2022, the verification subcall
```

File: src/theory/quantifiers/sygus/synth_verify.h [hunk 2]
```diff
@@ -56,6 +56,12 @@ class SynthVerify : protected EnvObj
...
+   * uses SMT not SMTO. Instead f is treated as an ordinary function symbol,
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/synth_verify.h:63]:
```suggestion
   * uses SMT, not SMTO. Instead f is treated as an ordinary function symbol,
```

File: src/theory/quantifiers/sygus/synth_verify.cpp [hunk 3]
```diff
@@ -111,12 +111,29 @@ Result SynthVerify::verify(Node query,
...
+            Assert(r.getStatus() == Result::UNKNOWN)
+                << "Expected model from verification step to satisfy query";
+          }
+          // If the query did not simplify to true, then it may be that the
+          // value for an oracle function was not what we expected.
+          if (options().quantifiers.oracles)
+          {
+            // In this case, we reconstruct the query, which may include more
+            // information about oracles than we had previously. We rerun the
+            // satisfiability check above.
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/synth_verify.cpp:129]: Clarify that this is (if I understand correctly) due to the I/O pairs found during search being added to the query.

Comment (ajreynol) [reply src/theory/quantifiers/sygus/synth_verify.cpp]: The I/O pairs are actually added in `TermDbSygus::rewriteNode`.  I've commented.

**Pull request merged**